File: daily_python/1514_Path_with_Maximum_Probability.py
import os
import sys
from typing import List
from collections import defaultdict
import heapq

class Solution:
    def maxProbability(self, n: int, edges: List[List[int]], succProb: List[float], start_node: int, end_node: int) -> float:
        e = defaultdict(list)
        for i, (u,v) in enumerate(edges):
            e[u].append((v, succProb[i]))
            e[v].append((u, succProb[i]))
        maxProb = [0.0] * n
        maxProb[start_node] = 1.0
        pq = []
        heapq.heappush(pq, (-1.0, start_node))
        round = 0
        while round < n and pq:
            curr_prob, curr_node = heapq.heappop(pq)
            if curr_node == end_node:
                return -curr_prob
            for neighbor, pathsucc in e[curr_node]:
                if maxProb[neighbor] < pathsucc * -curr_prob:
                    maxProb[neighbor] = pathsucc * -curr_prob
                    heapq.heappush(pq, (-maxProb[neighbor], neighbor))
        return 0.0

def readDataSet(filename):
    dataset = []
    with open(filename, 'r') as file:
        content = file.read()
        blocks = content.split('\n\n')
        for block in blocks:
            lines = block.split('\n')
            n = int(lines[0].split('=')[1].strip()[:-1])
            pairs = lines[1].split('=')[1][3:-3].split('},{')
            edges = [list(map(int, pair.split(','))) for pair in pairs]
            succProb = list(map(float, lines[2].split('=')[1][2:-2].split(',')))
            start_node = int(lines[3].split('=')[1].strip()[:-1])
            end_node = int(lines[4].split('=')[1].strip()[:-1])
            dataset.append((n, edges, succProb, start_node, end_node))
    return dataset

if __name__ == '__main__':
    if (len(sys.argv) == 1):
        filename = os.path.splitext(__file__)[0] + '.txt'
    else:
        filename = sys.argv[1]
    dataset = readDataSet(filename)
    results = []
    solution = Solution()
    for n, edges, succProb, start_node, end_node in dataset:
        results.append(solution.maxProbability(n, edges, succProb, start_node, end_node))
    for index, result in enumerate(results):
        print(f'Example {index + 1} : {result}')

File: daily/2024/1514_Path_with_Maximum_Probability.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

class Solution {
public:
    struct Node{
        int destination;
        double probability;
        Node* next;
        Node(int d, double p, Node* n) : destination(d), probability(p), next(n) {}
    };
    double maxProbability(int n, vector<vector<int>>& edges, vector<double>& succProb, int start_node, int end_node) {
        vector<double> prob(n, 0);
        prob[start_node] = 1.0;
        vector<Node*> e(n);
        for (int i = 0; i < edges.size(); i++){
            int n1 = edges[i][0];
            int n2 = edges[i][1];
            double p = succProb[i];
            e[n1] = new Node(n2, p, e[n1]);
            e[n2] = new Node(n1, p, e[n2]);
        }
        priority_queue<pair<double, int>> pq;
        pq.push({1.0, start_node});
        int round = 0;
        while(!pq.empty() && round < n){
            round++;
            auto [p, node] = pq.top();
            if (node == end_node) return p;
            pq.pop();
            Node* neighbor = e[node];
            while(neighbor){
                if (prob[neighbor->destination] < p * neighbor->probability){
                    prob[neighbor->destination] = p * neighbor->probability;
                    pq.push({prob[neighbor->destination], neighbor->destination});
                }
                neighbor = neighbor->next;
            }
        }
        return prob[end_node];
    }
};

int main(){
    int n = 3;
    vector<vector<int>> edges = {{0,1},{1,2},{0,2}};
    vector<double> succProb = {0.5,0.5,0.2};
    int start_node = 0;
    int end_node = 2;
    // int n = 3;
    // vector<vector<int>> edges = {{0,1},{1,2},{0,2}};
    // vector<double> succProb = {0.5,0.5,0.3};
    // int start_node = 0;
    // int end_node = 2;
    // int n = 3;
    // vector<vector<int>> edges = {{0,1}};
    // vector<double> succProb = {0.5};
    // int start_node = 0;
    // int end_node = 2;
    Solution S;
    cout << S.maxProbability(n, edges, succProb, start_node, end_node) << endl;
    return 0;
}